### PR TITLE
SY-4139: Schematic QoL Improvements

### PR DESCRIPTION
**Pull request by sy-nico**
# Issue Pull Request

## Linear Issue

[SY-4139](https://linear.app/synnax/issue/SY-4139/schematic-qol-improvements)

## Description

This PR introduces some low-lift Schematic QoL improvements, including frequently requested symbol grouping:

- Add context menu with `cut`/`copy`/`paste`
- Enable keyboard short-cut`cut` (ctrl+v)
- Introduce schematic grouping symbol
  - No nesting (for simplicity, can be implemented in the future)
  - Functions properly against multi-symbol alignment operations 
  - Context menu items added
  - Toolbar menu items added
  - Keyboard shortcut support
- Reorder Toolbar elements and improve dynamic wrapping.

https://github.com/user-attachments/assets/1c8f698b-78e7-46ec-a425-524790b54597

## Basic Readiness

- [x] I have performed a self-review of my code.
- [x] I have added relevant, automated tests to cover the changes.

<h3>Greptile Summary</h3>

This PR adds schematic grouping (symbol group container, group/ungroup via keyboard, toolbar, and context menu), a full cut/copy/paste context menu with keyboard shortcuts, and group-aware alignment/distribution in the Properties toolbar.

- **Grouping**: new `groupBox` node type, `useGroup`/`useUngroup` hooks, `Ctrl+G`/`Ctrl+U` triggers, group propagation on drag, and a post-dispatch `auditGroups` pass that dissolves single-member containers and resizes bounding boxes.
- **Clipboard**: refactored into pure `buildPayload`/`buildPasteActions`/`buildCutActions` helpers with imperative `copy`/`cut`/`paste` fns using `navigator.clipboard` (MIME + `text/plain`), resolving the prior cross-path incompatibility.
- **Alignment**: `resolveAlignmentKey` and `expandGroupPositions` ensure selected group members are deduplicated and their container is moved together during align/distribute operations.

<h3>Confidence Score: 4/5</h3>

Safe to merge with one visual regression fixed: the groupBox spec must use zIndex -1 before palette-added containers will layer correctly.

The groupBox spec sets zIndex 0, but the programmatic group-creation path in useGroup explicitly sets zIndex -1 so the container renders behind its members. Because groupBox is now registered in GROUP.variants, a user can drag one from the symbol palette and get a container that floats in front of member nodes. All other logic — group drag propagation, cascade removal, clipboard round-trips, and alignment deduplication — is well-covered by new unit tests and appears correct.

pluto/src/schematic/node/general/groupBox/external.tsx — the zIndex field needs to be -1 to match the programmatic creation path in queries.ts.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| pluto/src/schematic/node/general/groupBox/external.tsx | New groupBox symbol spec; zIndex: 0 is inconsistent with useGroup's zIndex: -1, causing palette-added containers to render above members. |
| pluto/src/schematic/groups.ts | New group-management utilities (expand/audit/propagate/align); propagateGroupDrag delta-capture guard uses zero-check instead of a dedicated flag, creating a subtle false-zero edge case. |
| pluto/src/schematic/clipboard.ts | Clipboard refactored into pure helpers; adds imperative copy/cut/paste using navigator.clipboard API with matching MIME + text/plain slots. |
| pluto/src/schematic/queries.ts | Adds useGroup, useUngroup, useSelectCanGroup, useSelectCanUngroup, useSelectAllConfigs; group/ungroup hooks dispatch via non-undoable dispatch (known limitation from prior review). |
| pluto/src/schematic/Schematic.tsx | Adds context menu (Cut/Copy/Paste/Group/Ungroup), Ctrl+G/Ctrl+U triggers, and group-aware handleNodesChange with cascade removal and post-dispatch auditGroups. |
| console/src/schematic/toolbar/Properties.tsx | Adds Group/Ungroup toolbar buttons and wires alignment/distribute operations through resolveAlignmentKey/expandGroupPositions so grouped symbols are treated as units. |
| pluto/src/schematic/groups.spec.ts | Comprehensive unit tests for all group utilities; good coverage of edge cases including empty inputs, cross-group isolation, and orphan detection. |
| pluto/src/schematic/clipboard.spec.tsx | Extends clipboard tests with group-aware copy/cut/paste, key remapping validation, and navigator.clipboard imperative API tests using FakeBlob/FakeClipboardItem stubs. |
| pluto/src/vis/diagram/Diagram.tsx | Adds onCut prop and handleCut handler mirroring the existing onCopy pattern; clean addition. |

</details>

<details><summary><h3>Comments Outside Diff (3)</h3></summary>

1. `console/src/schematic/Schematic.tsx`, line 438-447 ([link](https://github.com/synnaxlabs/synnax/blob/b02d9028bca41d6187af1cef3377388f01b45496/console/src/schematic/Schematic.tsx#L438-L447)) 

   <a href="#">P2</a> **Paste is not undoable**

   `handlePasteSelection` dispatches via the plain `dispatch`, not `undoableDispatch`, so pasting nodes cannot be reversed with `Ctrl+Z`. This is inconsistent with `handleCutSelection` (line 434) which correctly uses `undoableDispatch`. Consider using `undoableDispatch` here to make paste undoable.

2. `console/src/schematic/slice.ts`, line 230-250 ([link](https://github.com/synnaxlabs/synnax/blob/fbb83b67de5fb4d8ad27e2bec54e801c16439fc7/console/src/schematic/slice.ts#L230-L250)) 

   <a href="#">P1</a> **Cut group member preserves stale `groupId` in paste buffer**

   When a user cuts a group member (a node with `groupId`) without cutting the group container, the copied node retains its `groupId` in the buffer. `remapGroupIds` only remaps a `groupId` if the old group key appears in `keyMap` — but since the group container wasn't cut, it has no entry there. When pasted into the same schematic the node silently re-joins its original group; when pasted into a different schematic the `groupId` becomes a dangling reference.

   Clear `groupId` on copied nodes whose group container is not also in the copy buffer by iterating `copyBuffer.props` after building it and deleting `groupId` for any entry whose `groupId` is not present in the set of copied node keys.

2. `console/src/schematic/slice.ts`, line 238 ([link](https://github.com/synnaxlabs/synnax/blob/40fe3f26d3bb57bedc778977affddc703aaf9465/console/src/schematic/slice.ts#L238)) 

   <a href="#">P1</a> **Props assigned by reference corrupts copy buffer on repeated paste**

   `schematic.props[key] = state.copy.props[node.key]` shares the same Immer draft proxy between the pasted node and the copy buffer. `remapGroupIds` called four lines later then mutates `schematic.props[key].groupId`, which also mutates `state.copy.props[node.key].groupId` through the shared proxy — corrupting the buffer in-place.

   On a second paste the buffer's `groupId` values point to the first paste's newly-created keys; `remapGroupIds` can't find them in the new `keyMap` and deletes them, so the group structure is silently stripped from every paste after the first.

   Fix: shallow-spread when copying to break the shared reference:

3. `pluto/src/schematic/clipboard.ts`, line 177-230 ([link](https://github.com/synnaxlabs/synnax/blob/9997aabf399fce32752b8abeb1278dd5b8205ff7/pluto/src/schematic/clipboard.ts#L177-L230)) 

   <a href="#">P1</a> **Keyboard and context-menu clipboard paths are mutually incompatible**

   `handleCopy`/`handleCut` (fired by `Ctrl+C`/`Ctrl+X`) write the JSON payload to the custom MIME type and write `describe(built.payload)` — a human-readable string like `"3 nodes, 1 edge"` — to `text/plain`. The `paste()` function (context-menu Paste) reads `navigator.clipboard.readText()` which returns the `text/plain` slot; `tryParsePayload("3 nodes, 1 edge")` throws a JSON parse error and returns `null`, so nothing is pasted.

   The reverse is equally broken: `copy()`/`cut()` (context-menu Copy/Cut) write raw JSON to `text/plain` via `navigator.clipboard.writeText`. When `Ctrl+V` fires, `handlePaste` calls `e.clipboardData.getData(MIME)` which returns `""` (the custom MIME slot was never populated), `tryParsePayload("")` returns `null`, and the paste is silently dropped.

   The fix is to write the JSON payload — not the human-readable description — to `text/plain` in `handleCopy`/`handleCut`, so that `paste()` can correctly read and parse it.
</details>

<sub>Reviews (11): Last reviewed commit: ["SY-4139: Fix cross-path clipboard and em..."](https://github.com/synnaxlabs/synnax/commit/be4f290de86625be77d546103fb5b2a018b5c4ed) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=30267549)</sub>

> Greptile also left **1 inline comment** on this PR.

### CODE REVIEW

File: console/src/arc/editor/graph/Editor.tsx [hunk 1]
```diff
@@ -258,6 +258,7 @@ export const Editor: Layout.Renderer = ({ layoutKey, visible }) => {
 
   Diagram.useTriggers({
     onCopy: handleCopySelection,
+    onCut: handleCopySelection,
```
[the remark this answers is not in the document]
Comment (sy-nico) [reply console/src/arc/editor/graph/Editor.tsx]: Intentional. `onCut` is required by `UseTriggersProps`, and the arc editor doesn't support cut yet. Copy-only is the correct no-op for now.

File: console/src/schematic/slice.ts [hunk 2]
```diff
@@ -163,43 +172,61 @@ export interface SetSelectedSymbolGroupPayload {
...
+    },
+    // Copies the selected nodes and edges into the copy buffer, then removes
+    // them from the schematic.
+    cutSelection: (state, { payload }: PayloadAction<CutSelectionPayload>) => {
+      copySelectedToBuffer(state);
+      const { key: layoutKey } = payload;
+      const schematic = state.schematics[layoutKey];
+      const { nodes, edges } = state.copy;
+      const nodeKeys = new Set(nodes.map((n) => n.key));
+      const edgeKeys = new Set(edges.map((e) => e.key));
```
[the remark this answers is not in the document]
Comment (sy-nico) [reply console/src/schematic/slice.ts]: Pre-existing behavior. `clearOtherSelections` ensures only one schematic has active selections at a time, so this is safe. Will clean up any way.

File: console/src/schematic/Schematic.tsx [hunk 3]
```diff
@@ -414,20 +423,32 @@ export const Loaded: Layout.Renderer = ({ layoutKey, visible }) => {
...
 
   const handleCopySelection = useCallback(
-    (cursor: xy.XY) =>
-      dispatch(copySelection({ pos: calculateCursorPosition(cursor) })),
-    [dispatch, calculateCursorPosition],
+    (_cursor: xy.XY) => {
+      if (!canEdit) return;
+      dispatch(copySelection({ key: layoutKey }));
+    },
+    [canEdit, dispatch, layoutKey],
```
[the remark this answers is not in the document]
Comment (sy-nico) [reply console/src/schematic/Schematic.tsx]: A view-only schematic is, by definition, view only. This should not have been permitted and likely not used. This was exposed when the context menu was added. It would not make sense for only the copy to be enabled while in a "view only" mode.

image

**Comment by sy-nico** [comment]
Force-pushed: replaced the head with the migrated implementation on the sy-3038
architecture (action-based undo/redo, Pluto-owned flux store).

- Base: `rc` → `sy-3038-add-server-side-undo-and-redo` (stacked PR; will move
  back to `rc` once #2306 merges)
- Head: `612478d9` → `9997aabf3`. Pre-migration head preserved locally as
  backup.
- Diff includes the sy-3038 refactor; effective SY-4139 diff will surface once
  #2306 lands.

Inline comments from prior review may now reference code that's been moved or
removed in the new architecture.

**Comment by sy-nico** [comment]
@greptile, this branch has rebased onto [sy-3038-add-server-side-undo-and-redo](https://github.com/synnaxlabs/synnax/tree/sy-3038-add-server-side-undo-and-redo) and the feature re-implemented post-refactor. Update your review. Importantly, remove any stale "Comments Outside Diff" (3) that are no longer relevant due to the changes.

File: pluto/src/schematic/clipboard.ts [hunk 4]
```diff
@@ -70,84 +181,84 @@ export const useClipboard = ({
...
-      if (raw === "") return;
-      let payload: Payload;
-      try {
-        payload = JSON.parse(raw) as Payload;
-      } catch {
-        return;
-      }
-      if (payload.version !== VERSION) return;
+      const payload = tryParsePayload(e.clipboardData.getData(MIME));
+      if (payload == null) return;
```
[the remark this answers is not in the document]
Comment (sy-nico) [reply pluto/src/schematic/clipboard.ts:243]: Fixed. `copy()` / `cut()` / `paste()` now use `navigator.clipboard.write([new ClipboardItem({...})])` and `navigator.clipboard.read()`, writing both `MIME` and `text/plain` in one call. `describe()` preserved. Restores the pattern from `d9c728ef08` on sy-3038.

**Comment by sy-nico** [comment]
@greptile,

1: Stale — `handlePasteSelection` and `undoableDispatch` no longer exist. Undo/redo migrated to action-based dispatch via sy-3038.
2: Stale — `remapGroupIds` lives in `pluto/src/schematic/groups.ts` now as `remapGroupId`, which already drops `groupId` when the container is absent from `keyMap`.
3: Stale — copy buffer no longer uses Immer; payload is JSON-serialized to the OS clipboard. No shared references possible.
4: Fixed — see `clipboard.ts` change to `ClipboardItem` + `navigator.clipboard.read()`.
